### PR TITLE
Fix failures in relative-color-out-of-gamut-expected due to rgbToHwb algorithm

### PR DESCRIPTION
**Pull request by squelart**
#### d26cb4ae7f44f61334632273de133bd18023b67d
<pre>
Fix failures in relative-color-out-of-gamut-expected due to rgbToHwb algorithm
<a href="https://bugs.webkit.org/show_bug.cgi?id=277538">https://bugs.webkit.org/show_bug.cgi?id=277538</a>
<a href="https://rdar.apple.com/133045692">rdar://133045692</a>

Reviewed by Tim Nguyen.

The suggested rgbToHwb algorithm at <a href="https://drafts.csswg.org/css-color-4/#rgb-to-hwb">https://drafts.csswg.org/css-color-4/#rgb-to-hwb</a>
used to use rgbToHsl to compute the hue, discarding the saturation and lightness.
But rgbToHsl special-cases negative saturations (possible with out-of-gamut colors),
and flips the hue in that case, resulting in incorrect out-of-gamut HWB values.
This was fixed in <a href="https://github.com/w3c/csswg-drafts/pull/10718">https://github.com/w3c/csswg-drafts/pull/10718</a>

This patch here replicates <a href="https://github.com/w3c/csswg-drafts/pull/10718">https://github.com/w3c/csswg-drafts/pull/10718</a> in the
equivalent WebKit function, it fixes failures in
relative-color-out-of-gamut-expected.html, but also in
color-mix-out-of-gamut-expected.html.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-mix-out-of-gamut-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/relative-color-out-of-gamut-expected.txt:
* Source/WebCore/platform/graphics/ColorConversion.cpp:
(WebCore::ExtendedSRGBA&lt;float&gt;&gt;::convert):

Canonical link: <a href="https://commits.webkit.org/282101@main">https://commits.webkit.org/282101@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/f6a1b271f5006ea5a93652ede059d1810957d37c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61854 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41208 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14446 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65858 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12399 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63973 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48894 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12670 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49856 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8584 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64923 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38267 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53573 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30687 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34911 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10806 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11330 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56720 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11110 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67562 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5797 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10839 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57228 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5822 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53519 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57470 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13798 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4753 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37008 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38092 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39188 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37837 "Built successfully") | | |

### CODE REVIEW

**Comment by squelart** [comment]
Note: This patch does not reverse (parts of) https://commits.webkit.org/277999@main, because I wrote it before I looked into a cause. Reviewer(s), let me know if you'd prefer me to actually try and reverse (assuming the whole course of action is the right way to go, of course).

**Comment by weinig** [comment]
Thanks for looking into this @squelart!

Is this fix purely based on what the current WPT tests are testing, or was there some conversion with the WG on this? 

Perhaps I am misunderstanding standing you, but it seems like you are saying WebKit is doing what the spec suggests. If that is the case, should the test be updated instead?

**Comment by squelart** [comment]
> Is this fix purely based on what the current WPT tests are testing, or was there some conversation with the WG on this?

Yes this PR intends to fix what the WPT tests are testing, but I think it's particular to WebKit because of how it handles `hwb(from lab(100 104.3 -50.9) h w b)` by closely following the specs (more below).

No conversation with WG, maybe we should start one, especially if the specs need updating.

> Perhaps I am misunderstanding you, but it seems like you are saying WebKit is doing what the spec suggests. If that is the case, should the test be updated instead?

tl;dr: I believe WebKit follows the specs very closely, and it results in out-of-gamut errors.

The first failing test is:
```
  fuzzy_test_computed_color(`hwb(from lab(100 104.3 -50.9) h w b)`, `color(srgb 1.59343 0.58802 1.40564)`);
```
Now we have similar tests above that are *not* failing, e.g.:
```
  fuzzy_test_computed_color(`rgb(from lab(0 104.3 -50.9) r g b)`, `color(srgb 0.351376 -0.213938 0.299501)`);
```
Notice that the "from" colors are the same, there is a straight colorspace conversion, and the resulting relative colors are compared against the same srgb color.
This tells me that the test is most probably correct, especially since the whole test is about out-of-gamut colors, I would assume we want these to be preserved.
I've also verified this conversion in other places (other browsers, apps, and websites with their own js implementation).

Now, WebKit handles `hwb(from lab(100 104.3 -50.9) h w b)` internally by first converting to RGB (correctly), and then to HWB. I don't believe the specs suggest this, this is an implementation detail inside WebKit.

In any case, as extra proof that this is the point of failure, I can add the following test line:
```
  fuzzy_test_computed_color(`hwb(from color(srgb 1.59343 0.58802 1.40564) h w b / alpha)`, `color(srgb 1.59343 0.58802 1.40564)`);
```
And this fails the same way, pointing at the rgb-to-hwb conversion.

Now about the specs, we have https://drafts.csswg.org/css-color-4/#rgb-to-hwb
> 8.2. Converting sRGB Colors to HWB
> 
> Conversion in the reverse direction proceeds similarly. _[from above: "Converting an HWB color to sRGB is straightforward, and related to how one converts HSL to RGB"]_
> [...]
```
function rgbToHwb(red, green, blue) {
    var hsl = rgbToHsl(red, green, blue);
    var white = Math.min(red, green, blue);
    var black = 1 - Math.max(red, green, blue);
    return([hsl[0], white*100, black*100]);
}
```
As you can see this uses `rgbToHsl` as-is, but then only uses the hue component.

`rgbToHsl` is defined at https://drafts.csswg.org/css-color-4/#rgb-to-hsl, and there it says:
> Special care is taken to deal with intermediate negative values of saturation, which can be produced by colors far outside the sRGB gamut.

With the corresponding bit of code:
```
    // Very out of gamut colors can produce negative saturation
    // If so, just rotate the hue by 180 and use a positive saturation
    // see https://github.com/w3c/csswg-drafts/issues/9222
    if (sat < 0) {
        hue += 180;
        sat = Math.abs(sat);
    }
```
As it says, the hue is rotated and the saturation is made positive. This is probably the correct thing to do for HSL colors; my naive thinking is that it's like normalizing a vector with negative length by flipping its direction and negating the length, the resulting vector is the same.
But remember that `rgbToHwb` above only keeps the hue and drops the saturation, and I believe that's where the error ultimately comes from, we've effectively lost information.
Hence my proposed solution here: `rgbToHwb` should only use the parts of `rgbToHsl` that compute the pure hue, without influence from saturation (nor lightness).

Would you like me to open an issue against the specs? My suggestion for a spec change would be similar to this PR: Rewrite the `rgbToHwb` function so that it doesn't use `rgbToHsl`, but instead it should just do the minimum hue computation.

Sam, do you agree with my analysis?
And if you agree that the tests are correct (as they match results from multiple sources), I think this WebKit-specific fix could be merged before any potential WG decisions.

**Comment by weinig** [comment]
I think the tests are wrong, given they are supposed to test the spec. 

Let’s take this to CSS WG GitHub for discussion. 

If we want to make a short term fix, I would propose updating the test to match the spec, and pushing that to WPT.

**Comment by squelart** [comment]
Alright, I've filed https://github.com/w3c/csswg-drafts/issues/10695
While writing that I've found another argument for fixing the specs: The round trip `hwbToRgb(rgbToHwb(rgb(...)))` is broken for some out-of-gamut colors, and my suggestion fixes that. See https://codepen.io/squelart/pen/MWMoeoB for PoC.
Anyway, we can argue over at CSS WG before coming back here. 😄

**Review by nt1m — APPROVED** [review]
Matches the spec change at: https://github.com/w3c/csswg-drafts/commit/ead8668d310a164a8beef5af615e061a3cee95be
